src/compositor/hwcimage.cpp

```diff
@@ -0,0 +1,543 @@
+
+        if (image.size().isValid()) {
+
+            QPainter p(&image);
+
+            // Apply overlay
+            if (overlay.isValid())
+                p.fillRect(image.rect(), overlay);
+
+            // Apply glass..
+            if (effect.contains(QStringLiteral("glass"))) {
+                QImage glass("//usr/share/themes/jolla-ambient/meegotouch/icons/graphic-shader-texture.png");
+                glass = glass.scaled(glass.width() * pixelRatio, glass.height() * pixelRatio,
+                                     Qt::IgnoreAspectRatio, Qt::SmoothTransformation);
+                p.save();
+                p.setOpacity(0.1);
+                p.setCompositionMode(QPainter::CompositionMode_Plus);
+                p.fillRect(image.rect(), glass);
+                p.restore();
+            }
+        }
+    }
+
+    void run() {
+        execute();
+        mutex.lock();
+        if (hwcImage && hwcImage->m_pendingRequest == this) {
+            hwcImage->m_pendingRequest = 0;
+            QCoreApplication::postEvent(hwcImage, this);
+        } else {
+            setAutoDelete(true);
+        }
+        mutex.unlock();
+    }
+
+    QImage image;
+    QString file;
+    QString effect;
+    QColor overlay;
+    QSize textureSize;
+    qreal pixelRatio;
+    qreal rotation;
+
+    HwcImage *hwcImage;
+
+    static QMutex mutex;
+};
+
+QMutex HwcImageLoadRequest::mutex;
+
+HwcImage::HwcImage()
+    : m_pendingRequest(0)
+    , m_rotationHandler(0)
+    , m_status(Null)
+    , m_textureRotation(0)
+    , m_asynchronous(true)
+{
+    setFlag(ItemHasContents, true);
+}
+
+HwcImage::~HwcImage()
+{
+    HwcImageLoadRequest::mutex.lock();
+    if (m_pendingRequest)
+        m_pendingRequest->hwcImage = 0;
+    HwcImageLoadRequest::mutex.unlock();
+}
+
+
+/* We want to track the item that does rotation changes explicitly. This isn't
+   strictly needed, but it is easy to do and means we get away with tracking
+   one single object as opposed to listening for changes in an entire parent
+   hierarchy which would also be a lot slower.
+  */
+void HwcImage::setRotationHandler(QQuickItem *item)
+{
+    if (m_rotationHandler == item)
+        return;
+    if (hwcimage_is_enabled() && m_rotationHandler)
+        disconnect(m_rotationHandler, &QQuickItem::rotationChanged, this, &HwcImage::handlerRotationChanged);
+    m_rotationHandler = item;
+    if (hwcimage_is_enabled() && m_rotationHandler)
+        connect(m_rotationHandler, &QQuickItem::rotationChanged, this, &HwcImage::handlerRotationChanged);
+    emit rotationHandlerChanged(m_rotationHandler);
+    polish();
+}
+
+void HwcImage::setAsynchronous(bool is)
+{
+    if (m_asynchronous == is)
+        return;
+    m_asynchronous = is;
+    emit asynchronousChanged(m_asynchronous);
+    polish();
+}
+
+void HwcImage::setSource(const QUrl &source)
+{
+    if (source == m_source)
+        return;
+    m_source = source;
+    emit sourceChanged(m_source);
```

Comment:
Change the status to Loading here, otherwise something like image.source = "..."; if (image.status == HwcImage.Ready) { /* do thing immediately */ } else { /* defer thing */ } will fail.